faunadb-java-dsl/src/main/java/com/faunadb/client/types/Encoder.java

```diff
@@ -0,0 +1,186 @@
+package com.faunadb.client.types;
+
+import com.faunadb.client.errors.FaunaException;
+import com.faunadb.client.types.Properties.Property;
+import com.faunadb.client.types.Value.*;
+import com.faunadb.client.types.time.HighPrecisionTime;
+import org.joda.time.Instant;
+import org.joda.time.LocalDate;
+
+import java.lang.reflect.Array;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Stack;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Lists.newArrayListWithCapacity;
+import static com.google.common.collect.Maps.newHashMap;
+import static java.lang.String.format;
+
+/**
+ * FaunaDB object to {@link Value} encoder.
+ */
+public final class Encoder {
+  private Encoder() {
+  }
+
+  private final static Stack<Object> stack = new Stack<>();
+  private final static ConcurrentHashMap<Class<?>, Property[]> READ_PROPERTIES = new ConcurrentHashMap<>();
+
+  /**
+   * Encode the specified object into a corresponding FaunaDB value.
+   * <p>
+   * <pre>{@code
+   * encode(10) => new LongV(10)
+   * encode(3.14) => new DoubleV(3.14)
+   * encode(true) => BooleanV.TRUE
+   * encode(null) => NullV.NULL
+   * encode("a string") => new StringV("a string")
+   * encode(new int[] {1, 2}) => new ArrayV(Arrays.asList(1, 2))
+   * encode(Arrays.asList(1, 2)) => new ArrayV(Arrays.asList(1, 2))
+   * encode(new byte[] {1, 2}) => new ByteV(new byte[] {1, 2})
+   * encode(user) => new ObjectV(ImmutableMap.of("user_name", new StringV("john"), "password", new StringV("s3cr3t")))
+   * }</pre>
+   *
+   * @param obj Any instance of user defined classes, primitive values or any
+   *            generic collection like {@link java.util.List}, {@link java.util.Set} or {@link java.util.Map}
+   * @return A FaunaDB {@link Value} corresponding to the given argument
+   */
+  public static Value encode(Object obj) {
+    if (obj == null)
+      return Value.NullV.NULL;
+
+    if (Value.class.isInstance(obj))
+      return (Value) obj;
+
+    if (stack.contains(obj))
+      throw new FaunaException(format("Self reference loop detected for object `%s`", obj));
+
+    try {
+      stack.push(obj);
+
+      return encodeIntern(obj);
+    } finally {
+      stack.pop();
+    }
+  }
+
+  private static Value encodeIntern(Object obj) {
+    Class<?> clazz = obj.getClass();
+
+    if (clazz.isEnum())
+      return Enums.getEncodingMap((Class<Enum>) clazz).get(obj);
+
+    if (clazz == String.class)
```

Comment:
We should try to make this part of the `Codec` interface. The fact that this is separated from codec makes it more error phone as the wire protocol changes and we have to maintain different code for serialization and deserialization.

IMO, the `Codec` interface should be refactored to look something like:

```java
public interface Codec<T> {
    Result<T> decode(Value value);
    Result<Value> encode(T value);

   // Constants with codecs implementation
}

That way, the `Decoder` and `Encoder` job is just to reflect on the data structures finding the proper codec to use.

I'm not sure but, perhaps we should consider to make a `CodecRegistry` where people can register their custom codecs extensibility. For example, if you're working with a library and you wan't their data structures persisted into Fauna. Idk... maybe that's something to think about.